cpp/ql/src/experimental/Security/CWE/CWE-190/AllocMultiplicationOverflow.ql

```diff
@@ -0,0 +1,42 @@
+/**
+ * @name Multiplication result may overflow and be used in allocation
+ * @description TODO
+ * @kind path-problem
+ * @problem.severity TODO
+ * @precision TODO
+ * @tags security
+ *       correctness
+ *       external/cwe/cwe-190
+ *       external/cwe/cwe-128
+ * @id cpp/multiplication-overflow-in-alloc
+ */
+
+import cpp
+import semmle.code.cpp.models.interfaces.Allocation
+import semmle.code.cpp.dataflow.DataFlow
+import DataFlow::PathGraph
+
+class MultToAllocConfig extends DataFlow::Configuration {
+  MultToAllocConfig() { this = "MultToAllocConfig" }
+
+  override predicate isSource(DataFlow::Node node) {
+    // a multiplication of two non-constant expressions
+    exists(MulExpr me |
+      me = node.asExpr() and
+      forall(Expr e | e = me.getAnOperand() | not exists(e.getValue()))
```

Comment:
This should be equivalent, and AFAIK could end up being optimized better.
```suggestion
      not exists(Expr e | e = me.getAnOperand() and exists(e.getValue()))
```

Reply:
What is the intuition for it being better optimized?

Reply:
Hm. Looking at my notes I see now that I was wrong - I think it doesn't make any difference when the body of the `forall` doesn't involve recursion, so feel free to choose whatever form you find most readable.

Reply:
Out of interest, what is the issue when the body involves recursion?  (assuming negative recursion doesn't crop up)

Reply:
It's an example of non-linear recursion, which is slow because the recursive call cannot just be compiled down to a join with the `prev_delta` of the previous iteration. It's briefly mentioned in https://docs.google.com/document/d/12LURUTA51n1xeuoyGHhnSNTTjIa-1LQFG2sb_i-WhNI/edit#.